File: rust/core/src/session.rs
//! High-level session management for World ID verification

use crate::{
    bridge::{BridgeClient, BridgeConfig, Status},
    types::{AppId, BridgeUrl, Credential, Proof, Request, VerificationLevel},
    Constraints, ConstraintNode, Error, Result,
};
use std::time::Duration;
use tokio::time::sleep;

/// Configuration for creating a World ID verification session
#[derive(Debug, Clone)]
pub struct SessionConfig {
    /// Application ID
    pub app_id: AppId,

    /// Action identifier
    pub action: String,

    /// Optional action description shown to users
    pub action_description: Option<String>,

    /// Credential requests
    pub requests: Vec<Request>,

    /// Optional constraints on which credentials are acceptable
    pub constraints: Option<Constraints>,

    /// Bridge URL (defaults to production)
    pub bridge_url: Option<BridgeUrl>,
}

impl SessionConfig {
    /// Creates a new session config with required fields
    #[must_use]
    pub fn new(app_id: AppId, action: impl Into<String>) -> Self {
        Self {
            app_id,
            action: action.into(),
            action_description: None,
            requests: Vec::new(),
            constraints: None,
            bridge_url: None,
        }
    }

    /// Adds an action description
    #[must_use]
    pub fn with_description(mut self, description: impl Into<String>) -> Self {
        self.action_description = Some(description.into());
        self
    }

    /// Adds a single request
    #[must_use]
    pub fn with_request(mut self, request: Request) -> Self {
        self.requests.push(request);
        self
    }

    /// Sets all requests
    #[must_use]
    pub fn with_requests(mut self, requests: Vec<Request>) -> Self {
        self.requests = requests;
        self
    }

    /// Sets constraints
    #[must_use]
    pub fn with_constraints(mut self, constraints: Constraints) -> Self {
        self.constraints = Some(constraints);
        self
    }

    /// Sets the bridge URL
    #[must_use]
    pub fn with_bridge_url(mut self, url: BridgeUrl) -> Self {
        self.bridge_url = Some(url);
        self
    }

    /// Creates a session config from a legacy verification level
    ///
    /// This provides backward compatibility with the old API
    #[must_use]
    pub fn from_verification_level(
        app_id: AppId,
        action: impl Into<String>,
        verification_level: VerificationLevel,
        signal: impl Into<String>,
    ) -> Self {
        let signal_str = signal.into();
        let credentials = verification_level.to_credentials();

        let requests = credentials
            .iter()
            .map(|cred| Request::new(*cred, crate::crypto::encode_signal_str(&signal_str)))
            .collect();

        let constraints = Constraints::new(ConstraintNode::any(
            credentials
                .into_iter()
                .map(ConstraintNode::credential)
                .collect(),
        ));

        Self {
            app_id,
            action: action.into(),
            action_description: None,
            requests,
            constraints: Some(constraints),
            bridge_url: None,
        }
    }
}

/// A World ID verification session
pub struct Session {
    client: BridgeClient,
}

impl Session {
    /// Creates a new session from configuration
    ///
    /// # Errors
    ///
    /// Returns an error if the session cannot be created
    pub async fn create(config: SessionConfig) -> Result<Self> {
        let bridge_config = BridgeConfig {
            app_id: config.app_id,
            action: config.action,
            action_description: config.action_description,
            requests: config.requests,
            constraints: config.constraints,
            bridge_url: config.bridge_url.unwrap_or_else(BridgeUrl::default),
        };

        let client = BridgeClient::create(bridge_config).await?;

        Ok(Self { client })
    }

    /// Returns the connect URL that should be presented to the user
    #[must_use]
    pub fn connect_url(&self) -> String {
        self.client.connect_url()
    }

    /// Polls for the current status once
    ///
    /// # Errors
    ///
    /// Returns an error if the poll request fails
    pub async fn poll(&self) -> Result<Status> {
        self.client.poll_status().await
    }

    /// Waits for a proof, polling the bridge until completion
    ///
    /// # Errors
    ///
    /// Returns an error if polling fails or the verification fails
    pub async fn wait_for_proof(&self) -> Result<Proof> {
        self.wait_for_proof_with_timeout(Duration::from_secs(300))
            .await
    }

    /// Waits for a proof with a timeout
    ///
    /// # Errors
    ///
    /// Returns an error if polling fails, verification fails, or timeout is reached
    pub async fn wait_for_proof_with_timeout(&self, timeout: Duration) -> Result<Proof> {
        let start = tokio::time::Instant::now();
        let poll_interval = Duration::from_secs(3);

        loop {
            if start.elapsed() > timeout {
                return Err(Error::Timeout);
            }

            match self.poll().await? {
                Status::Confirmed(proof) => return Ok(proof),
                Status::Failed(error) => return Err(Error::AppError(error)),
                Status::WaitingForConnection | Status::AwaitingConfirmation => {
                    sleep(poll_interval).await;
                }
            }
        }
    }

    /// Returns the request ID for this session
    #[must_use]
    pub fn request_id(&self) -> String {
        self.client.request_id().to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_session_config_builder() {
        let app_id = AppId::new("app_test").unwrap();
        let config = SessionConfig::new(app_id.clone(), "test_action")
            .with_description("Test description")
            .with_request(Request::new(
                Credential::Orb,
                crate::crypto::encode_signal_str("test"),
            ))
            .with_constraints(Constraints::any(vec![Credential::Orb]));

        assert_eq!(config.action, "test_action");
        assert_eq!(config.action_description, Some("Test description".to_string()));
        assert_eq!(config.requests.len(), 1);
        assert!(config.constraints.is_some());
    }

    #[test]
    fn test_session_config_from_verification_level() {
        let app_id = AppId::new("app_test").unwrap();
        let config = SessionConfig::from_verification_level(
            app_id,
            "test_action",
            VerificationLevel::Device,
            "test_signal",
        );

        assert_eq!(config.requests.len(), 2); // Orb and Device
        assert!(config.constraints.is_some());
    }
}

File: js/packages/core/src/session.ts
/**
 * IDKit Session
 * Pure functional API for World ID verification - no dependencies
 */

import type {
  VerifyConfig,
  ConstraintNode,
  CredentialType,
  CredentialRequestType,
  RpContext,
} from "./types/config";
import type { ISuccessResult } from "./types/result";
import { AppErrorCodes } from "./types/bridge";
import { WasmModule, initIDKit } from "./lib/wasm";

/** Options for pollForUpdates() */
export interface WaitOptions {
  /** Milliseconds between polls (default: 1000) */
  pollInterval?: number;
  /** Total timeout in milliseconds (default: 300000 = 5 minutes) */
  timeout?: number;
  /** AbortSignal for cancellation */
  signal?: AbortSignal;
}

/** Status returned from pollOnce() */
export interface Status {
  type:
    | "waiting_for_connection"
    | "awaiting_confirmation"
    | "confirmed"
    | "failed";
  proof?: ISuccessResult;
  error?: AppErrorCodes;
}

// Re-export RpContext for convenience
export type { RpContext };

/**
 * A World ID verification session
 *
 * Provides a clean, promise-based API for World ID verification flows.
 * Each session represents a single verification attempt.
 */
export interface Session {
  /** QR code URL for World App - display this as a QR code for users to scan */
  readonly connectorURI: string;
  /** Unique request ID for this verification */
  readonly requestId: string;
  /** Poll once for current status (for manual polling) */
  pollOnce(): Promise<Status>;
  /** Poll continuously until completion or timeout */
  pollForUpdates(options?: WaitOptions): Promise<ISuccessResult>;
}

/**
 * Internal session implementation
 */
class SessionImpl implements Session {
  private wasmSession: WasmModule.Session;
  private _connectorURI: string;
  private _requestId: string;

  constructor(wasmSession: WasmModule.Session) {
    this.wasmSession = wasmSession;
    this._connectorURI = wasmSession.connectUrl();
    this._requestId = wasmSession.requestId();
  }

  get connectorURI(): string {
    return this._connectorURI;
  }

  get requestId(): string {
    return this._requestId;
  }

  async pollOnce(): Promise<Status> {
    return (await this.wasmSession.pollForStatus()) as Status;
  }

  async pollForUpdates(options?: WaitOptions): Promise<ISuccessResult> {
    const pollInterval = options?.pollInterval ?? 1000;
    const timeout = options?.timeout ?? 300000; // 5 minutes default
    const startTime = Date.now();

    while (true) {
      // Check for cancellation
      if (options?.signal?.aborted) {
        throw new Error("Verification cancelled");
      }

      // Check timeout
      if (Date.now() - startTime > timeout) {
        throw new Error(`Timeout waiting for proof after ${timeout}ms`);
      }

      // Poll status
      const status = await this.pollOnce();

      if (status.type === "confirmed" && status.proof) {
        return status.proof;
      }

      if (status.type === "failed") {
        const errorCode = status.error ?? AppErrorCodes.GenericError;
        throw new Error(`Verification failed: ${errorCode}`);
      }

      // Wait before next poll
      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// CredentialRequest and Constraint helpers
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Creates a CredentialRequest for a credential type
 *
 * @param credential_type - The type of credential to request (e.g., 'orb', 'face')
 * @param options - Optional signal and genesis_issued_at_min
 * @returns A CredentialRequest object
 *
 * @example
 * ```typescript
 * const orb = CredentialRequest('orb', { signal: 'user-123' })
 * const face = CredentialRequest('face')
 * ```
 */
export function CredentialRequest(
  credential_type: CredentialType,
  options?: { signal?: string; genesis_issued_at_min?: number },
): CredentialRequestType {
  return {
    type: credential_type,
    signal: options?.signal,
    genesis_issued_at_min: options?.genesis_issued_at_min,
  };
}

/**
 * Creates an OR constraint - at least one child must be satisfied
 *
 * @param nodes - Constraint nodes (CredentialRequests or nested constraints)
 * @returns An "any" constraint node
 *
 * @example
 * ```typescript
 * const constraint = any(CredentialRequest('orb'), CredentialRequest('face'))
 * ```
 */
export function any(...nodes: ConstraintNode[]): { any: ConstraintNode[] } {
  return { any: nodes };
}

/**
 * Creates an AND constraint - all children must be satisfied
 *
 * @param nodes - Constraint nodes (CredentialRequests or nested constraints)
 * @returns An "all" constraint node
 *
 * @example
 * ```typescript
 * const constraint = all(CredentialRequest('orb'), any(CredentialRequest('document'), CredentialRequest('secure_document')))
 * ```
 */
export function all(...nodes: ConstraintNode[]): { all: ConstraintNode[] } {
  return { all: nodes };
}

// ─────────────────────────────────────────────────────────────────────────────
// Preset helpers
// ─────────────────────────────────────────────────────────────────────────────

/**
 * OrbLegacy preset configuration
 */
export interface OrbLegacyPreset {
  type: "OrbLegacy";
  data: { signal?: string };
}

/**
 * Preset types for simplified session creation
 */
export type Preset = OrbLegacyPreset;

/**
 * Creates an OrbLegacy preset for World ID 3.0 legacy support
 *
 * This preset creates a session compatible with both World ID 4.0 and 3.0 protocols.
 * Use this when you need backward compatibility with older World App versions.
 *
 * @param opts - Optional configuration with signal
 * @returns An OrbLegacy preset
 *
 * @example
 * ```typescript
 * const session = await verify({ app_id, action, rp_context })
 *   .preset(orbLegacy({ signal: 'user-123' }))
 * ```
 */
export function orbLegacy(opts: { signal?: string } = {}): OrbLegacyPreset {
  return { type: "OrbLegacy", data: { signal: opts.signal } };
}

// ─────────────────────────────────────────────────────────────────────────────
// VerifyBuilder
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Builder for creating verification sessions
 */
class VerifyBuilder {
  private config: VerifyConfig;

  constructor(config: VerifyConfig) {
    this.config = config;
  }

  /**
   * Creates a verification session with the given constraints
   *
   * @param constraints - Constraint tree (CredentialRequest or any/all combinators)
   * @returns A new Session instance
   *
   * @example
   * ```typescript
   * const session = await verify({ app_id, action, rp_context })
   *   .constraints(any(CredentialRequest('orb'), CredentialRequest('face')))
   * ```
   */
  async constraints(constraints: ConstraintNode): Promise<Session> {
    // Ensure WASM is initialized
    await initIDKit();

    // Create WASM RpContext
    const rpContext = new WasmModule.RpContextWasm(
      this.config.rp_context.rp_id,
      this.config.rp_context.nonce,
      BigInt(this.config.rp_context.created_at),
      BigInt(this.config.rp_context.expires_at),
      this.config.rp_context.signature,
    );

    // Create WASM VerifyBuilder and call constraints
    const wasmBuilder = WasmModule.verify(
      this.config.app_id,
      String(this.config.action),
      rpContext,
      this.config.action_description ?? null,
      this.config.bridge_url ?? null,
    );

    const wasmSession = (await wasmBuilder.constraints(
      constraints,
    )) as unknown as WasmModule.Session;

    return new SessionImpl(wasmSession);
  }

  /**
   * Creates a verification session from a preset
   *
   * Presets provide a simplified way to create sessions with predefined
   * credential configurations. The preset is converted to both World ID 4.0
   * constraints and World ID 3.0 legacy fields for backward compatibility.
   *
   * @param preset - A preset object from orbLegacy()
   * @returns A new Session instance
   *
   * @example
   * ```typescript
   * const session = await verify({ app_id, action, rp_context })
   *   .preset(orbLegacy({ signal: 'user-123' }))
   * ```
   */
  async preset(preset: Preset): Promise<Session> {
    // Ensure WASM is initialized
    await initIDKit();

    // Create WASM RpContext
    const rpContext = new WasmModule.RpContextWasm(
      this.config.rp_context.rp_id,
      this.config.rp_context.nonce,
      BigInt(this.config.rp_context.created_at),
      BigInt(this.config.rp_context.expires_at),
      this.config.rp_context.signature,
    );

    // Create WASM VerifyBuilder and call preset
    const wasmBuilder = WasmModule.verify(
      this.config.app_id,
      String(this.config.action),
      rpContext,
      this.config.action_description ?? null,
      this.config.bridge_url ?? null,
    );

    const wasmSession = (await wasmBuilder.preset(
      preset,
    )) as unknown as WasmModule.Session;

    return new SessionImpl(wasmSession);
  }
}

/**
 * Creates a verification builder
 *
 * This is the main entry point for creating World ID verification sessions.
 * Use the builder pattern with constraints to specify which credentials to accept.
 *
 * @param config - Verification configuration
 * @returns A VerifyBuilder instance
 *
 * @example
 * ```typescript
 * import { verify, CredentialRequest, any, initIDKit } from '@worldcoin/idkit-core'
 *
 * // Initialize WASM (only needed once)
 * await initIDKit()
 *
 * // Create request items
 * const orb = CredentialRequest('orb', { signal: 'user-123' })
 * const face = CredentialRequest('face')
 *
 * // Create a verification session with constraints
 * const session = await verify({
 *   app_id: 'app_staging_xxxxx',
 *   action: 'my-action',
 *   rp_context: {
 *     rp_id: 'rp_123456789abcdef0',
 *     nonce: 'unique-nonce',
 *     created_at: Math.floor(Date.now() / 1000),
 *     expires_at: Math.floor(Date.now() / 1000) + 3600,
 *     signature: 'ecdsa-signature-from-backend',
 *   },
 * }).constraints(any(orb, face))
 *
 * // Display QR code
 * console.log('Scan this:', session.connectorURI)
 *
 * // Wait for proof
 * const proof = await session.pollForUpdates()
 * console.log('Success:', proof)
 * ```
 */
export function verify(config: VerifyConfig): VerifyBuilder {
  // Validate required fields
  if (!config.app_id) {
    throw new Error("app_id is required");
  }
  if (!config.action) {
    throw new Error("action is required");
  }
  if (!config.rp_context) {
    throw new Error("rp_context is required");
  }

  return new VerifyBuilder(config);
}

